File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) {

        if(grid == null || grid.length == 0)
            return 0;


        int count = 0;
        for(int i=0;i<grid.length;i++) {
            for(int j=0;j<grid[i].length;j++) {
                // 
                if(grid[i][j] == '1') {
                    count = count + bfs(grid, i, j);
                }
            }
        }

        return count;
    }

    public int bfs(char[][] grid, int i, int j) {
        if(Math.min(i,j) < 0 || i == grid.length || j == grid[0].length || grid[i][j] == '0') {
            return 0;
        }

        grid[i][j] = '0';

        bfs(grid, i+1, j);
        bfs(grid, i-1, j);
        bfs(grid, i, j+1);
        bfs(grid, i, j-1);

        return 1;
    }
}

File: Data Structures & Algorithms/shortest-path-in-binary-matrix/submission-0.py
from collections import deque
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        ROW, COL = len(grid), len(grid[0])
        if grid[0][0] or grid[ROW-1][COL-1]:
            return -1

        length = 1
        queue = deque()
        queue.append((0,0, length))
        visited = set()
        visited.add((0,0))

        
        

        while queue:
            size = len(queue)
            for i in range(size):

                r,c, length = queue.popleft()
                if r == ROW-1 and c == COL-1:
                    return length

                directions = [(0, 1), (1, 0), (0, -1), (-1, 0),
                  (1, 1), (-1, -1), (1, -1), (-1, 1)]

                for dr,dc in directions:
                    nr, nc = r+dr, c+dc
                    if min(nr,nc) < 0 or nr == ROW or nc == COL or (nr,nc) in visited or grid[nr][nc] == 1:
                        continue
                    
                    print((nr,nc))
                    queue.append((nr,nc, length+1))
                    visited.add((nr,nc))

        return -1


                


        
        

        